src/main/java/com/amazon/aws/iot/greengrass/semver/SemVer.java

```diff
@@ -0,0 +1,340 @@
+
+    /**
+     * Get the normalized value of the SemVer including prerelease and build
+     * @return String.
+     */
+    public String getValue() {
+        if (build.isEmpty()) {
+            return version;
+        }
+        return String.format("%s+%s", version, String.join(".", build));
+    }
+
+    @Override
+    public int compareTo(SemVer other) {
+        if (other.version.equals(this.version)) {
+            return 0;
+        }
+
+        int mainCompared = this.compareMain(other);
+        if (mainCompared != 0) {
+            return mainCompared;
+        }
+        return this.comparePre(other);
+    }
+
+    public int compareMain(SemVer other) {
+        int maj = compareIdentifiers(this.major, other.major);
+        if (maj != 0) {
+            return maj;
+        }
+        int min = compareIdentifiers(this.minor, other.minor);
+        if (min != 0) {
+            return min;
+        }
+        return compareIdentifiers(this.patch, other.patch);
+    }
+
+    public int comparePre(SemVer other) {
+        // NOT having a prerelease is > having one
+        if (!this.prerelease.isEmpty() && other.prerelease.isEmpty()) {
+            return -1;
+        } else if (this.prerelease.isEmpty() && !other.prerelease.isEmpty()) {
+            return 1;
+        } else if (this.prerelease.isEmpty() && other.prerelease.isEmpty()) {
+            return 0;
+        }
+
+        for (int i = 0; ; i++) {
+            if (i > this.prerelease.size() - 1 && i > other.prerelease.size() - 1) {
+                return 0;
+            } else if (i > other.prerelease.size() - 1) {
+                return 1;
+            } else if (i > this.prerelease.size() - 1) {
+                return -1;
+            }
+            Object a = this.prerelease.get(i);
+            Object b = other.prerelease.get(i);
+            if (Objects.equals(a, b)) {
+                continue;
+            }
+            return compareIdentifiers(Objects.toString(a), Objects.toString(b));
+        }
+    }
+
+    public int compareBuild(SemVer other) {
+        for (int i = 0; ; i++) {
+            if (i > this.build.size() - 1 && i > other.build.size() - 1) {
+                return 0;
+            } else if (i > other.build.size() - 1) {
+                return 1;
+            } else if (i > this.build.size() - 1) {
+                return -1;
+            }
+            String a = this.build.get(i);
+            String b = other.build.get(i);
+            if (Objects.equals(a, b)) {
+                continue;
+            } else {
+                return compareIdentifiers(a, b);
+            }
+        }
+    }
+
+    static Integer isInt(String x) {
+        try {
+            return Integer.parseInt(x);
+        } catch (NumberFormatException f) {
+            return null;
+        }
+    }
+
+    public static int compareIdentifiers(int a, int b) {
+        return Integer.compare(a, b);
+    }
+
+    public static int compareIdentifiers(String a, String b) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.